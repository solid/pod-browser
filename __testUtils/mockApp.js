/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addStringNoLocale,
  addUrl,
  createSolidDataset,
  createThing,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { vcard, rdf, schema } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { vcardExtras } from "../src/addressBook";

export const oidcPrefix = "http://www.w3.org/ns/solid/oidc#";
export const TOS_PREDICATE = `${oidcPrefix}tos_uri`;
export const POLICY_PREDICATE = `${oidcPrefix}policy_uri`;
export const CONTACTS_PREDICATE = `${oidcPrefix}contacts`;
export const LOGO_PREDICATE = `${oidcPrefix}logo_uri`;

export const APP_WEBID = "https://mockappurl.com";
export const APP_POLICY_URL = "https://mockappurl.com/privacy-policy";
export const APP_TOS_URL = "https://mockappurl.com/terms-of-service";

const contactThing = chain(
  createThing({ url: "https://mockappurl.com#contacts" }),
  (t) => addStringNoLocale(t, vcard.fn, "Example Contact"),
  (t) => addStringNoLocale(t, vcard.hasTelephone, "555-5555"),
  (t) => addStringNoLocale(t, vcard.hasEmail, "example@email.com"),
  (t) => addUrl(t, vcard.url, "https://examplewebsite.com")
);

const dataset = setThing(createSolidDataset(), contactThing);

export function mockApp() {
  return chain(
    createThing({ url: APP_WEBID }),
    (t) => addStringNoLocale(t, vcard.fn, "Mock App"),
    (t) => addUrl(t, vcardExtras("WebId"), APP_WEBID),
    (t) => addUrl(t, rdf.type, schema.SoftwareApplication), // just guessing what the type might be for apps
    (t) => addUrl(t, TOS_PREDICATE, APP_TOS_URL),
    (t) => addUrl(t, POLICY_PREDICATE, APP_POLICY_URL),
    (t) => addUrl(t, CONTACTS_PREDICATE, "https://mockappurl.com#contacts")
  );
}

export function mockAppDataset() {
  return setThing(dataset, mockApp());
}
