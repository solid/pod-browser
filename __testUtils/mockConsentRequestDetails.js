/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

async function getConsentRequestDetails() {
  return {
    "@context": [
      "https://www.w3.org/2018/credentials/v1",
      "https://w3id.org/security/suites/ed25519-2020/v1",
      "https://consent.pod.inrupt.com/credentials/v1",
    ],
    id: "https://consent.pod.inrupt.com/vc/53973727-f4d0-9e8dbdc041fd",
    type: ["VerifiableCredential", "SolidCredential", "SolidConsentRequest"],
    issuer: "https://consent.pod.inrupt.com",
    issuanceDate: "2021-05-26T16:40:03Z",
    expirationDate: "2021-06-23T16:40:03Z",
    credentialSubject: {
      id: "https://pod.inrupt.com/bob/profile/card#me",
      inbox: "https://pod.inrupt.com/bob/inbox/",
      hasConsent: {
        mode: ["Read", "Write"],
        hasStatus: "ConsentStatusRequested",
        forPersonalData: "https://pod.inrupt.com/alice/private/data",
        forPurpose: "https://example.com/SomeSpecificPurpose",
      },
    },
    proof: {
      created: "2021-05-26T16:40:03.009Z",
      proofPurpose: "assertionMethod",
      proofValue: "eqp8h_kL1DwJCpn65z-d1Arnysx6b11...jb8j0MxUCc1uDQ",
      type: "Ed25519Signature2020",
      verificationMethod: "https://consent.pod.inrupt.com/key/396f686b4ec",
    },
  };
}

export default getConsentRequestDetails;
