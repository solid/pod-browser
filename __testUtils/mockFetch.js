/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

export default function mockFetch(responses = {}) {
  return jest.fn().mockImplementation((url, init) => {
    const responseFn = responses[url.toString()];
    const responseObject = responseFn;
    const method = init?.method || "GET";
    let response;
    if (responseFn && typeof responseObject[method] !== "undefined") {
      response = responseObject[method](url, init);
    } else if (responseFn) {
      response = responseFn(url, init);
    } else {
      throw new Error(`URL (${url}) not mocked properly`);
    }
    Object.defineProperty(response, "url", { value: url });
    return Promise.resolve(response);
  });
}

export function mockSolidClientRequest(body, method = "GET", headers = {}) {
  return new Request(body, {
    headers: new Headers(headers),
    method,
  });
}

export function mockTurtleResponse(status, body = "") {
  return new Response(body, {
    status,
    headers: {
      "Content-Type": "text/turtle",
    },
  });
}
