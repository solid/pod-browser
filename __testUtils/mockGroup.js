/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import {
  addUrl,
  asUrl,
  getThing,
  mockSolidDatasetFrom,
  mockThingFrom,
  removeUrl,
  setStringNoLocale,
  setThing,
  setUrl,
} from "@inrupt/solid-client";
import { foaf, rdf, vcard } from "rdf-namespaces";
import { chain } from "../src/solidClientHelpers/utils";
import { getBaseUrl } from "../src/solidClientHelpers/resource";

export function addMembersToMockedGroup(group, agentUrls) {
  const dataset = chain(
    group.dataset,
    ...agentUrls.map((agentUrl) => (d) =>
      setThing(d, setUrl(group.thing, vcard.hasMember, agentUrl))
    )
  );

  return {
    dataset,
    thing: getThing(dataset, asUrl(group.thing)),
  };
}

export function removeMembersFromMockedGroup(group, agentUrls) {
  const dataset = chain(
    group.dataset,
    ...agentUrls.map((agentUrl) => (d) =>
      setThing(d, removeUrl(group.thing, vcard.hasMember, agentUrl))
    )
  );

  return {
    dataset,
    thing: getThing(dataset, asUrl(group.thing)),
  };
}

export function mockGroupThing(name, url, { description, members } = {}) {
  return chain(
    mockThingFrom(url),
    (t) => setUrl(t, rdf.type, vcard.Group),
    (t) => setStringNoLocale(t, foaf.name, name),
    ...(members || []).map((agentUrl) => (t) =>
      addUrl(t, vcard.hasMember, agentUrl)
    ),
    ...[
      description ? (t) => setStringNoLocale(t, vcard.note, description) : null,
    ].filter((fn) => !!fn)
  );
}

export default function mockGroup(name, url, options = {}) {
  const thing = mockGroupThing(name, url, options);
  const dataset = setThing(mockSolidDatasetFrom(getBaseUrl(url)), thing);
  return { dataset, thing };
}
