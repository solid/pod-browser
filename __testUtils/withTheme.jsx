/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { render } from "@testing-library/react";
import T from "prop-types";
import { StylesProvider, ThemeProvider } from "@inrupt/prism-react-components";
import { FeatureProvider } from "../src/contexts/featureFlagsContext";
import defaultTheme from "../src/theme";

export default function WithTheme(props) {
  const { theme, children } = props;
  return (
    <StylesProvider generateClassName={(rule) => `PodBrowser-${rule.key}`}>
      <ThemeProvider theme={theme}>{children}</ThemeProvider>
    </StylesProvider>
  );
}

WithTheme.propTypes = {
  // eslint-disable-next-line react/forbid-prop-types
  theme: T.object,
  children: T.node.isRequired,
};

WithTheme.defaultProps = {
  theme: defaultTheme,
};

export const renderWithTheme = (children, theme = defaultTheme) => {
  return render(
    <FeatureProvider>
      <WithTheme theme={theme}>{children}</WithTheme>
    </FeatureProvider>
  );
};
