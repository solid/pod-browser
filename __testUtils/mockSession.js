/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import mockFetch from "./mockFetch";
import mockResponse from "./mockResponse";

import profileTurtle from "./mocks/profile.ttl";
import storageTurtle from "./mocks/podRoot.ttl";
import storageAclTurtle from "./mocks/storageAcl.ttl";

export const webIdUrl = "http://example.com/webId#me";
export const storageUrl = "http://example.com/";
export const storageAclUrl = "http://example.com/.acl";
export const anotherUsersStorageUrl = "http://anotheruser.com/";
export const anotherUsersStorageAclUrl = "http://anotheruser.com/.acl";

export { profileTurtle };

export default function mockSession(options = {}) {
  // this instance is a normal, authenticated session
  // you can override the properties with what you need, but if you repeat something a lot
  // consider naming it and create a reusable mock function
  const { webId, ...restOptions } = options;
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webId || webIdUrl]: () =>
        mockResponse(200, profileTurtle, { "Content-Type": "text/turtle" }),
      [storageUrl]: () =>
        mockResponse(200, storageTurtle, {
          Link: `<${storageAclUrl}>; rel="acl"`,
          "Content-Type": "text/turtle",
        }),
      [storageAclUrl]: () =>
        mockResponse(200, storageAclTurtle, { "Content-Type": "text/turtle" }),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: options.webId || webIdUrl,
    }, // add more properties as needed
    ...restOptions,
  }; // add more properties as needed
}

// create other functions in this file to reflect other types of sessions we need to test with

export function mockUnauthenticatedSession() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: () => mockFetch(),
    info: {
      isLoggedIn: false,
      sessionId: null,
      webId: null,
    },
    login: jest.fn(),
  };
}

export function mockAuthenticatedSessionWithNoAccessToPod() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webIdUrl]: () => mockResponse(200, profileTurtle),
      [storageUrl]: () =>
        mockResponse(200, "", {
          Link: `<${storageAclUrl}>; rel="acl"`,
        }),
      [storageAclUrl]: () => mockResponse(401),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: webIdUrl,
    },
  };
}

export function mockAuthenticatedSessionWithNoAccessToAnotherUsersPod() {
  return {
    on: jest.fn(),
    handleIncomingRedirect: jest.fn().mockResolvedValue(null),
    fetch: mockFetch({
      [webIdUrl]: () => mockResponse(200, profileTurtle),
      [anotherUsersStorageUrl]: () =>
        mockResponse(200, "", {
          Link: `<${anotherUsersStorageAclUrl}>; rel="acl"`,
        }),
      [anotherUsersStorageAclUrl]: () => mockResponse(401),
    }),
    info: {
      isLoggedIn: true,
      sessionId: "some-session-id",
      webId: webIdUrl,
    },
  };
}

export const mockAuthenticatedSession =
  mockAuthenticatedSessionWithNoAccessToPod;
