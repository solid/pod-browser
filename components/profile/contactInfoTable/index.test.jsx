/**
 * Copyright 2020 Inrupt Inc.
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal in
 * the Software without restriction, including without limitation the rights to use,
 * copy, modify, merge, publish, distribute, sublicense, and/or sell copies of the
 * Software, and to permit persons to whom the Software is furnished to do so,
 * subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED,
 * INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A
 * PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT
 * HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
 * OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE
 * SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 */

import React from "react";
import { rdf, vcard } from "rdf-namespaces";
import * as scFns from "@inrupt/solid-client";
import { CombinedDataProvider, ThingProvider } from "@inrupt/solid-ui-react";
import { addUrl, getUrl, setThing } from "@inrupt/solid-client";
import { render } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import mockSession from "../../../__testUtils/mockSession";
import mockSessionContextProvider from "../../../__testUtils/mockSessionContextProvider";
import { renderWithTheme } from "../../../__testUtils/withTheme";
import ContactInfoTable, {
  CONTACT_INFO_TYPE_EMAIL,
  DEFAULT_CONTACT_TYPE,
  PREFIX_MAP,
  setupAddContactDetail,
  setupColumnTypeBody,
  setupColumnValueBody,
  setupDeleteButtonCell,
  setupOnChange,
  setupOnSave,
  setupRemoveRow,
  setupRowProps,
  setupSaveHandler,
} from "./index";
import useDataset from "../../../src/hooks/useDataset";
import {
  aliceWebIdUrl,
  mockPersonDatasetAliceWithContactInfo,
  mockPersonThingAliceWithContactInfo,
} from "../../../__testUtils/mockPersonResource";

jest.mock("../../../src/hooks/useDataset");
const mockedUseDataset = useDataset;

const { mockSolidDatasetFrom, mockThingFrom } = scFns;
const dataset = mockPersonDatasetAliceWithContactInfo();
const profile = mockThingFrom("http://example.com/profile#this");

beforeEach(() => {
  mockedUseDataset.mockReturnValue(dataset);
});

describe("ContactInfoTable", () => {
  it("renders a table of contact info", async () => {
    const session = mockSession({ fetch });
    const SessionProvider = mockSessionContextProvider(session);
    const contactInfo = [
      {
        type: vcard.Home,
        value: "tel:42-1337",
      },
      { type: vcard.Home, value: "mailto:alice@example.com" },
    ];

    const { asFragment, findByRole } = renderWithTheme(
      <SessionProvider>
        <ContactInfoTable
          property={vcard.hasEmail}
          contactInfoType={CONTACT_INFO_TYPE_EMAIL}
          values={contactInfo}
          webId={aliceWebIdUrl}
        />
      </SessionProvider>
    );
    await expect(findByRole("table")).resolves.not.toBeNull();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders an editable table of contact info", () => {
    const session = mockSession({ fetch });
    const SessionProvider = mockSessionContextProvider(session);
    const profileDataset = mockPersonDatasetAliceWithContactInfo();
    const thing = mockPersonThingAliceWithContactInfo();

    const { asFragment } = renderWithTheme(
      <SessionProvider>
        <CombinedDataProvider solidDataset={profileDataset} thing={thing}>
          <ContactInfoTable
            property={vcard.hasEmail}
            editing
            contactInfoType={CONTACT_INFO_TYPE_EMAIL}
            webId={aliceWebIdUrl}
          />
        </CombinedDataProvider>
      </SessionProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe("setupAddContactDetail", () => {
  it("handles adding telephone numbers", async () => {
    const property = vcard.hasTelephone;
    const contactType = DEFAULT_CONTACT_TYPE;
    const contactValue = "42-1337";
    const saveHandler = jest.fn();
    const setNewContactValue = jest.fn();

    await expect(
      setupAddContactDetail(
        property,
        contactType,
        contactValue,
        dataset,
        profile,
        saveHandler,
        setNewContactValue
      )()
    ).resolves.toBeUndefined();

    expect(saveHandler).toHaveBeenCalled();
    const newDataset = saveHandler.mock.calls[0][1];
    const newContactDetail = scFns.getThing(
      newDataset,
      "http://alice.example.com/phone"
    );
    expect(getUrl(newContactDetail, rdf.type)).toEqual(contactType);
    expect(getUrl(newContactDetail, vcard.value)).toEqual(
      `${PREFIX_MAP[property]}${contactValue}`
    );

    expect(setNewContactValue).toHaveBeenCalledWith("");
  });
});

describe("setupSaveHandler", () => {
  it("sets a save handler using saveSolidDatasetAt", async () => {
    const savedDataset = "savedDataset";
    jest.spyOn(scFns, "saveSolidDatasetAt").mockResolvedValue(savedDataset);
    const fetch = jest.fn();
    const setDataset = jest.fn();
    const newThing = mockThingFrom("http://example.com/#newThing");
    const datasetUrl = "http://example.com";
    const datasetToUpdate = mockSolidDatasetFrom(datasetUrl);

    await expect(
      setupSaveHandler(fetch, setDataset)(newThing, datasetToUpdate)
    ).resolves.toBeUndefined();
    expect(scFns.saveSolidDatasetAt).toHaveBeenCalledWith(
      datasetUrl,
      setThing(datasetToUpdate, newThing),
      { fetch }
    );
    expect(setDataset).toHaveBeenCalledWith(savedDataset);
  });
});

describe("setupRemoveRow", () => {
  it("removes details and saves change", async () => {
    const property = vcard.hasTelephone;
    const contactValue = "tel:test";
    const profileWithNumber = addUrl(profile, property, contactValue);
    const saveHandler = jest.fn();
    await expect(
      setupRemoveRow(
        profileWithNumber,
        property,
        saveHandler,
        dataset
      )(mockThingFrom(contactValue))
    ).resolves.toBeUndefined();
    expect(saveHandler).toHaveBeenCalled();
    const [newProfile] = saveHandler.mock.calls[0];
    expect(getUrl(newProfile, property)).toBeNull();
  });
});

describe("setupDeleteButtonCell", () => {
  const thing = mockThingFrom("http://example.com#thing");

  it("returns a button if editable", () => {
    const removeRow = jest.fn();
    const contactInfoType = CONTACT_INFO_TYPE_EMAIL;
    const Button = setupDeleteButtonCell(true, contactInfoType, removeRow);

    const { asFragment, container } = render(
      <ThingProvider thing={thing}>
        <Button />
      </ThingProvider>
    );

    expect(asFragment()).toMatchSnapshot();

    userEvent.click(container.querySelector("button"));
    expect(removeRow).toHaveBeenCalledWith(thing);
  });

  it("renders nothing if not editable", () => {
    const contactInfoType = CONTACT_INFO_TYPE_EMAIL;
    const Component = setupDeleteButtonCell(
      false,
      contactInfoType,
      () => {},
      () => {}
    );

    const { asFragment } = render(
      <ThingProvider thing={thing}>
        <Component />
      </ThingProvider>
    );

    expect(asFragment()).toMatchSnapshot();
  });
});

describe("setupRowProps", () => {
  it("sets up row props", () => {
    const bem = (value) => value;
    expect(setupRowProps(bem)()).toEqual({
      className: "table__body-row",
    });
  });
});

describe("setupColumnTypeBody", () => {
  it("sets up body for type", () => {
    const columnTypeBody = setupColumnTypeBody(CONTACT_INFO_TYPE_EMAIL);
    const { asFragment } = render(columnTypeBody({ value: vcard.Home }));
    expect(asFragment()).toMatchSnapshot();
  });

  it("has a fallback if value is not valid", () => {
    const columnTypeBody = setupColumnTypeBody(CONTACT_INFO_TYPE_EMAIL);
    const { asFragment } = render(columnTypeBody({ value: "invalid" }));
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("setupOnSave", () => {
  it("sets dataset", () => {
    const setDataset = jest.fn();
    expect(setupOnSave(setDataset)(dataset)).toBeUndefined();
    expect(setDataset).toHaveBeenCalledWith(dataset);
  });
});

describe("setupColumnValueBody", () => {
  const bem = (value) => value;
  const onSave = () => {};

  it("renders for read-only mode", () => {
    const { asFragment } = render(setupColumnValueBody(false, bem, onSave)());
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders for editing mode", () => {
    const { asFragment } = render(setupColumnValueBody(true, bem, onSave)());
    expect(asFragment()).toMatchSnapshot();
  });
});

describe("setupOnChange", () => {
  it("setups an onChange handler", () => {
    const setValue = jest.fn();
    const onChange = setupOnChange(setValue);
    const value = "value";
    expect(onChange({ target: { value } })).toBeUndefined();
    expect(setValue).toHaveBeenCalledWith(value);
  });
});
