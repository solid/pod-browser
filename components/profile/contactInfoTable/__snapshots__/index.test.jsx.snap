// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ContactInfoTable renders a table of contact info 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-root"
  >
    <table
      class="PodBrowser-table  PodBrowser-table--inherits PodBrowser-table"
    >
      <tbody>
        <tr>
          <td>
            <p
              class="PodBrowser-root PodBrowser-body1"
            >
              Home
            </p>
          </td>
          <td>
            <p
              class="PodBrowser-root PodBrowser-body1"
            >
              tel:42-1337
            </p>
          </td>
        </tr>
        <tr>
          <td>
            <p
              class="PodBrowser-root PodBrowser-body1"
            >
              Home
            </p>
          </td>
          <td>
            <p
              class="PodBrowser-root PodBrowser-body1"
            >
              mailto:alice@example.com
            </p>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;

exports[`ContactInfoTable renders an editable table of contact info 1`] = `
<DocumentFragment>
  <div
    class="PodBrowser-root PodBrowser-root"
  >
    <table
      class="PodBrowser-table  PodBrowser-table--inherits PodBrowser-table"
      role="table"
    >
      <thead>
        <tr
          role="row"
        >
          <th
            colspan="1"
            role="columnheader"
          />
          <th
            colspan="1"
            role="columnheader"
          />
          <th
            colspan="1"
            role="columnheader"
          />
        </tr>
      </thead>
      <tbody
        role="rowgroup"
      >
        <tr
          class="PodBrowser-table__body-row"
          role="row"
        >
          <td
            role="cell"
          >
            <p
              class="PodBrowser-root PodBrowser-body1"
              data-testid="profile-email-existing-type"
            />
          </td>
          <td
            role="cell"
          >
            <p
              class="PodBrowser-root PodBrowser-body1"
            >
              <input
                class="PodBrowser-input"
                data-testid="profile-email-existing-value"
                type="url"
                value="mailto:alice@example.com"
              />
            </p>
          </td>
          <td
            role="cell"
          >
            <button
              class="PodBrowser-button PodBrowser-button--secondary"
              data-testid="profile-email-delete-button"
            >
              <span
                class="PodBrowser-button__text"
              >
                Delete
              </span>
            </button>
          </td>
        </tr>
      </tbody>
    </table>
  </div>
  <div
    class="PodBrowser-root PodBrowser-root"
  >
    <div
      class="PodBrowser-root PodBrowser-root"
    >
      <div
        class="PodBrowser-root"
      >
        <div
          class="PodBrowser-root PodBrowser-root PodBrowser-formControl"
          data-testid="profile-email-new-type"
        >
          <div
            aria-haspopup="listbox"
            class="PodBrowser-root PodBrowser-select PodBrowser-selectMenu PodBrowser-outlined PodBrowser-input PodBrowser-input"
            role="button"
            tabindex="0"
          >
            Home
          </div>
          <input
            aria-hidden="true"
            class="PodBrowser-nativeInput"
            tabindex="-1"
            value="http://www.w3.org/2006/vcard/ns#Home"
          />
          <svg
            aria-hidden="true"
            class="PodBrowser-root PodBrowser-icon PodBrowser-iconOutlined"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M7 10l5 5 5-5z"
            />
          </svg>
          <fieldset
            aria-hidden="true"
            class="PodBrowser-root PodBrowser-notchedOutline"
            style="padding-left: 8px;"
          >
            <legend
              class="PodBrowser-legend"
              style="width: 0.01px;"
            >
              <span>
                â€‹
              </span>
            </legend>
          </fieldset>
        </div>
      </div>
    </div>
    <div
      class="PodBrowser-root PodBrowser-root"
    >
      <input
        class="PodBrowser-input"
        data-testid="profile-email-new-value"
        label="Value"
        value=""
      />
    </div>
    <div
      class="PodBrowser-root PodBrowser-root"
    >
      <button
        class="PodBrowser-button PodBrowser-button--secondary"
        data-testid="profile-email-add-button"
      >
        <span
          class="PodBrowser-button__text"
        >
          Add
        </span>
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`setupColumnTypeBody has a fallback if value is not valid 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1"
    data-testid="profile-email-existing-type"
    title="invalid"
  >
    invalid
  </p>
</DocumentFragment>
`;

exports[`setupColumnTypeBody sets up body for type 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1"
    data-testid="profile-email-existing-type"
    title="http://www.w3.org/2006/vcard/ns#Home"
  >
    Home
  </p>
</DocumentFragment>
`;

exports[`setupColumnValueBody renders for editing mode 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    <span>
      Error: Thing not found as property or in context
    </span>
  </p>
</DocumentFragment>
`;

exports[`setupColumnValueBody renders for read-only mode 1`] = `
<DocumentFragment>
  <p
    class="MuiTypography-root MuiTypography-body1"
  >
    <span>
      Error: Thing not found as property or in context
    </span>
  </p>
</DocumentFragment>
`;

exports[`setupDeleteButtonCell renders nothing if not editable 1`] = `<DocumentFragment />`;

exports[`setupDeleteButtonCell returns a button if editable 1`] = `
<DocumentFragment>
  <button
    class="makeStyles-button-1553 makeStyles-button--secondary-1559"
    data-testid="profile-email-delete-button"
  >
    <span
      class="makeStyles-button__text-1557"
    >
      Delete
    </span>
  </button>
</DocumentFragment>
`;
